couchdb/tests/server/qs.py

```diff
@@ -0,0 +1,583 @@
+# -*- coding: utf-8 -*-
+#
+import unittest
+
+from functools import partial
+from io import StringIO
+
+from couchdb import json
+from couchdb.server import BaseQueryServer, SimpleQueryServer
+from couchdb.server import exceptions
+from couchdb.server.helpers import wrap_func_to_ddoc
+
+
+class BaseQueryServerTestCase(unittest.TestCase):
+
+    def test_set_version(self):
+        server = BaseQueryServer((1, 2, 3))
+        self.assertEqual(server.version, (1, 2, 3))
+
+    def test_set_latest_version_by_default(self):
+        server = BaseQueryServer()
+        self.assertEqual(server.version, (999, 999, 999))
+
+    def test_set_config_option(self):
+        server = BaseQueryServer(foo='bar')
+        self.assertTrue('foo' in server.config)
+        self.assertEqual(server.config['foo'], 'bar')
+
+    def test_config_option_handler(self):
+        class CustomServer(BaseQueryServer):
+            def config_foo(self, value):
+                self.config['baz'] = value
+        server = CustomServer(foo='bar')
+        self.assertTrue('foo' not in server.config)
+        self.assertTrue('baz' in server.config)
+        self.assertEqual(server.config['baz'], 'bar')
+
+    def test_pass_server_instance_to_command_handler(self):
+        server = BaseQueryServer()
+        server.commands['foo'] = lambda s, x: server is s
+        self.assertTrue(server.process_request(['foo', 'bar']))
+
+    def test_raise_fatal_error_on_unknown_command(self):
+        server = BaseQueryServer(output=StringIO())
+        try:
+            server.process_request(['foo', 'bar'])
+        except Exception as err:
+            self.assertTrue(isinstance(err, exceptions.FatalError))
+            self.assertEqual(err.args[0], 'unknown_command')
+
+    def test_handle_fatal_error(self):
+        def command_foo(*a, **k):
+            raise exceptions.FatalError('foo', 'bar')
+
+        def maybe_fatal_error(func):
+            def wrapper(exc_type, exc_value, exc_traceback):
+                assert exc_type is exceptions.FatalError
+                return func(exc_type, exc_value, exc_traceback)
+            return wrapper
+
+        output = StringIO()
+        server = BaseQueryServer(output=output)
+        server.handle_fatal_error = maybe_fatal_error(server.handle_fatal_error)
+        server.commands['foo'] = command_foo
+        try:
+            server.process_request(['foo', 'bar'])
+        except Exception as err:
+            self.assertTrue(isinstance(err, exceptions.FatalError))
+
+    def test_response_for_fatal_error_oldstyle(self):
+        def command_foo(*a, **k):
+            raise exceptions.FatalError('foo', 'bar')
+
+        output = StringIO()
+        server = BaseQueryServer(version=(0, 9, 0), output=output)
+        server.commands['foo'] = command_foo
+        expected = {'reason': 'bar', 'error': 'foo'}
+        try:
+            server.process_request(['foo', 'bar'])
+        except Exception:
+            pass
```

Comment:
copy-paste I guess, that's why >_> 
the test goal was about response data, not the exception kind.